components/generic_modal.tsx

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import classNames from 'classnames';
+import React from 'react';
+import {Modal} from 'react-bootstrap';
+import {FormattedMessage} from 'react-intl';
+
+import './generic_modal.scss';
+
+type Props = {
+    onHide: () => void;
+    modalHeaderText: string | JSX.Element;
+    show?: boolean;
+    handleCancel?: () => void;
+    handleConfirm?: () => void;
+    confirmButtonText?: string | JSX.Element;
+    confirmButtonClassName?: string;
+    cancelButtonText?: string | JSX.Element;
+    isConfirmDisabled?: boolean;
+};
+
+type State = {
+    show: boolean;
+}
+
+export default class GenericModal extends React.PureComponent<Props, State> {
+    static defaultProps: Partial<Props> = {
+        show: true,
+    };
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            show: props.show!,
+        };
+    }
+
+    onHide = () => {
+        this.setState({show: false}, this.props.onHide);
+    }
+
+    handleCancel = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+        event.preventDefault();
+        this.onHide();
+        if (this.props.handleCancel) {
+            this.props.handleCancel();
+        }
+    }
+
+    handleConfirm = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+        event.preventDefault();
+        this.onHide();
+        if (this.props.handleConfirm) {
+            this.props.handleConfirm();
+        }
+    }
+
+    render() {
+        let confirmButton;
+        if (this.props.handleConfirm) {
+            let confirmButtonText = (
+                <FormattedMessage
+                    id='generic_modal.confirm'
+                    defaultMessage='Confirm'
+                />
+            );
+            if (this.props.confirmButtonText) {
+                confirmButtonText = (
+                    <React.Fragment>
```

Comment:
Continuing [this discussion](https://github.com/mattermost/mattermost-webapp/pull/5711#discussion_r443620657), I did a little googling and it looks like `React.ReactNode` is the preferred type for `confirmButtonText`. I guess it's automatically typed `JSX.Element` because of how it's set above, but `React.ReactNode` supports both cases.

Reply:
That said, I'd be fine leaving it as-is now that I understand the reasoning for this. I still wanted to call it out though since it's good to know the distinction between the two since we'll need to use `React.ReactNode` to type props like `children` elsewhere.

Reply:
I feel like I should know that, considering I just used that for cloud onboarding :P
I'll make the change.